lib/puppet/agent.rb

```diff
@@ -60,8 +61,22 @@ def run(client_options = {})
               end
             end
           rescue Puppet::LockError
-            Puppet.notice _("Run of %{client_class} already in progress; skipping  (%{lockfile_path} exists)") % { client_class: client_class, lockfile_path: lockfile_path }
-            return
+            wait_for_lock_deadline ||= Time.now.to_i + Puppet[:maxwaitforlock]
+
+            if Puppet[:waitforlock]  < 1
+              Puppet.notice _("Run of %{client_class} already in progress; skipping  (%{lockfile_path} exists)") % { client_class: client_class, lockfile_path: lockfile_path }
+              return 1
+            elsif Time.now.to_i > wait_for_lock_deadline
+              Puppet.notice _("Exiting now because the maxwaitforlock timeout has been exceeded.")
+              return 1
+            else
+              while File.file?(lockfile_path) && Time.now.to_i <= wait_for_lock_deadline
+                Puppet.info _("Another puppet instance is already running; --waitforlock flag used, waiting for running instance to finish.")
+                Puppet.info(_("Will try again in %{time} seconds.") % {time: Puppet[:waitforlock]})
+                sleep Puppet[:waitforlock]
+              end
+              retry
```

Comment:
It would be good to test the three cases, so if the lock is held by another instance then:

1. If `maxwaitforlock == 0` (default), then the `already in progress` message is printed
2. If `maxwaitforlock > 0`, then the `Exiting now..." message is printed 
3. If `maxwaitforlock > 0`, but the deadline hasn't been exceeded, then it sleeps and prints the "Will try again" message is printed, until it finally exits with "Exiting now".